src/server/ua_services_nodemanagement.c

```diff
@@ -1777,17 +1740,62 @@ recursiveDeleteNode(UA_Server *server, UA_Session *session,
         if(!child)
             continue;
         /* Only delete child nodes that have no other parent */
-        if(!multipleHierarchicalRefs(&child->head, hierarchRefsSet))
-            recursiveDeleteNode(server, session, hierarchRefsSet, &child->head, true);
+        if(!multipleHierarchicalRefs(&child->head, hierarchRefsSet, refTree)){
+            UA_Boolean contained;
+            UA_StatusCode addResult = RefTree_addNodeId(refTree, &child->head.nodeId, &contained);
+            if(addResult != UA_STATUSCODE_GOOD){
+                UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                            "Bad add result!");
+                return;
+            }
+        }
         UA_NODESTORE_RELEASE(server, child);
     }
-
     UA_BrowseResult_clear(&br);
+}
 
-    if(removeTargetRefs)
-        removeIncomingReferences(server, session, head);
+static void
+findSetOfSingleRefChilds(UA_Server *server, UA_Session *session,
+                    const UA_ReferenceTypeSet *hierarchRefsSet,
+                    const UA_NodeHead *head, UA_Boolean removeTargetRefs, RefTree *refTree) {
+    UA_StatusCode retval = RefTree_init(refTree);
```

Comment:
@jpfr The initialisiation is done here. But i can move it to the RefTree creation place to make it more visible.